### ISSUE TITLE
TEXTNOW SIP Credential

### ISSUE THREAD

**Issue by gronget** [gornostay25/TextNowSipJsPhone#1]
Hey
I was wondering if the textnow sip creds can be used in asterisk (as an external phone)

**Comment by thetechdudemc** [comment]
No, TextNow does not accept connections from a standard SIP client like a desktop softphone or Asterisk. I've been trying to do the same thing, and used the SIP credentials from here to try and connect and it doesn't work, it states the connection just times out. That's because TextNow only accepts connections from WebRTC and WebSockets. 

It's possible, and it would be really nice if someone could make a local proxy type program that accepts a standard SIP connection, then forward it to TextNow's servers using the required WebRTC and WebSocket protocol, but it's unlikely as it's a niche use case and people just pay a site like voip.ms instead

**Comment by st1905** [comment]
Asterisk has pjsip websocket transport support and i believe textnow is running pjsip. I remember this "REGISTER User-Agent: TextNow PjSip". So, you won`t get registered if you use it without websocket transport in Asterisk, I will try to get this done soon using openwrt.

**Comment by rasterburn1** [comment]
hey there, I know this thread is a year old now but as I was searching out some enlightenment towards an anomaly or what ever you want to call it within my own configuration I stumbled upon this thread and felt compelled to chime in. Let me just say this, I cant speak in regards to Asterisk specifically, but I can confirm form first hand experience (though very much heavily unsupported in the sense your on yourown from the get go) that you can connect to TextNow using your sip credentials to both make and receive calls through 3CX. 

I have had a 3CX setup (stock self hosted) functioning as a V.O. endpoint for TextNow since late June 2022. As far as calling and texting goesI havent had any issues with calling and still need the app or the web app in order to send or receive a text but for calling you still have the call limitations of TextNow's sip trunk with the added features of 3CX.

Prior to April 2022 (when I initially set up my 3CX system) I was considering jumping ship with TextNow and getting a new (business) number elsewhere due to the issues I was encountering along with the limitations of their system such as localized history/logging where the call or text will show up on the device it was received on but not on another (eg webapp, android app, iphone app), this problem has had me chasing my tail in circles for hours to track down a text message or call from someone a couple months later as only the device it was received on would show any signs of it ever existing. If I had missed a call from someone hopefully it was important enough that they would try calling again later because if they left a voicemail, 90% of the time I couldnt listen to it anyways cause it would just comeback with "encountered an error", that is if the call or voicemail would even show up. And then of course, on occation missing an inbound call because the app or phone wanted to wait a few moments from pressing the answer button to answering the call as well as missing a non-ringing call only to be notified 8hrs later that I missed the call(without leaving wifi and phone in pocket). not to mention all the onslaught of adverts that goes towards paying their costs to help keep the service free...ish  

From April to June 2022 everything was ready to go to jump ship from TextNow minus a new number and a simple method to redirect callers/texters to the new number or atleast notify them of the new number so I set forth and patched 3CX and TextNow together as a temporary solution to aid in the transition over to a new number.

June 2022 to now (as of jan.2023), that TextNow number is still configured as a SIP Trunk on 3CX and gets used mainly as a secondary outbound trunk and when signing up to things on the internet that request a phone number but shouldnt need one.

The pros I have noticed to this kind of setup:
1) TextNow's LIMITED options, and features are less applicable to you.
2) You gain the benefit of a centralized call log/history in 3CX (answered or not) and stable vm for missed calls 
3) options and features of a PBX become available such as digital receptionist, call recording, multi-user, so on so forth
4) Contacts phonebook that is visable and accessable regardless of device (I have yet to see one in the textnow webapp other then the shotty log)
5) Im sure there are a pile more benefits I havent noticed.

The cons I have noticed to this kind of setup:
1) Texting is unavailable and still requires a TextNow phone app or their webapp
2) Setup and configuration requirements and limitations are unknown
	- My setup consists of 3CX Self-Hosted Enterprise Edition (it was a free upgrade) installed on a Google Cloud e2-micro server with a "Lock-In" TextNow number (purchased to remove the effects of outbound inactivity)
3) 1 TextNow SIP Trunk account can be connected at a time (or at least from what I have found during my testing)
4) Various other limitations or requirements I have yet to notice.

In Short, Hell yeah you can utilize your TextNow Credentials via 3rd party PBX/voIP software.

**Comment by st1905** [comment]
Can you please write how you connected 3cx to textnow ? Any hints on your setup ? Thank you.

**Comment by rasterburn1** [comment]
um yeah, not the best at doing such but yeah, maybe able to give it a go lol

Most of this will probably seem ass backwards or face palm worthy but i assure you, it wont make alot of sense and it probably isnt the obvious things to be looking for
Step 1) log into your textnow account via a web browser and open up your devtools / debugging console
Step 2) look for lines that start with something like "DEBUG - phoneService", "INFO - sip.UserAgent", or "REGISTER sip:prod.tncp.textnow.com SIP/2.0", as these lines contain your SIP connection details including username, realm, uri, so on so forth
Step 3) make note of the super critical information
    INFO - sip.Registerer - · registrar: sip:prod.tncp.textnow.com:5070
    INFO - sip.Transport - WebSocket opened wss://web1.prod.tncp.textnow.com:443
    WWW-Authenticate: Digest realm="prod.tncp.textnow.com", nonce="??"
    Authorization: Digest algorithm=MD5, username="??", realm="prod.tncp.textnow.com", nonce="??", uri="sip:prod.tncp.textnow.com", response="??"
Step 4) set up a generic sip trunk and fill in the blanks

I have supplied a template with most of the generic data filled in, templated off my setup this morning so if you have 3CX, the template should be pretty plug and play and shouldnt be too difficult to alter for the use of another system...
[TextNow.pv.xml.txt](https://github.com/gornostay25/TextNowSipJsPhone/files/10460852/TextNow.pv.xml.txt)

Like I mentioned, I am not sure if there are any specific requirements or limitations towards the successfull connection, things such as requiring the "Lock-In Number" subscription or the "Preimier Number" subscription vs the basic free number as well as a limitation that only 1 will be able to connect at a time (though I have my doubts)

Also Like I mentioned, I have a "Lock-In Number" subscription which is connected just fine, while any attempt to connect a free number (only tested with the first connected) has failed to connect with the error "401 Unauthorized"

Well, hope that helps

**Comment by st1905** [comment]
Thank you, i will try with raspberry pi. I have asterisk on openwrt and currently using sip trunks with it. (They work quite good, having only some minor issues with DTMF)

**Comment by khyrfrj** [comment]
> hey there, I know this thread is a year old now but as I was searching out some enlightenment towards an anomaly or what ever you want to call it within my own configuration I stumbled upon this thread and felt compelled to chime in. Let me just say this, I cant speak in regards to Asterisk specifically, but I can confirm form first hand experience (though very much heavily unsupported in the sense your on yourown from the get go) that you can connect to TextNow using your sip credentials to both make and receive calls through 3CX.
> 
> I have had a 3CX setup (stock self hosted) functioning as a V.O. endpoint for TextNow since late June 2022. As far as calling and texting goesI havent had any issues with calling and still need the app or the web app in order to send or receive a text but for calling you still have the call limitations of TextNow's sip trunk with the added features of 3CX.
> 
> Prior to April 2022 (when I initially set up my 3CX system) I was considering jumping ship with TextNow and getting a new (business) number elsewhere due to the issues I was encountering along with the limitations of their system such as localized history/logging where the call or text will show up on the device it was received on but not on another (eg webapp, android app, iphone app), this problem has had me chasing my tail in circles for hours to track down a text message or call from someone a couple months later as only the device it was received on would show any signs of it ever existing. If I had missed a call from someone hopefully it was important enough that they would try calling again later because if they left a voicemail, 90% of the time I couldnt listen to it anyways cause it would just comeback with "encountered an error", that is if the call or voicemail would even show up. And then of course, on occation missing an inbound call because the app or phone wanted to wait a few moments from pressing the answer button to answering the call as well as missing a non-ringing call only to be notified 8hrs later that I missed the call(without leaving wifi and phone in pocket). not to mention all the onslaught of adverts that goes towards paying their costs to help keep the service free...ish
> 
> From April to June 2022 everything was ready to go to jump ship from TextNow minus a new number and a simple method to redirect callers/texters to the new number or atleast notify them of the new number so I set forth and patched 3CX and TextNow together as a temporary solution to aid in the transition over to a new number.
> 
> June 2022 to now (as of jan.2023), that TextNow number is still configured as a SIP Trunk on 3CX and gets used mainly as a secondary outbound trunk and when signing up to things on the internet that request a phone number but shouldnt need one.
> 
> The pros I have noticed to this kind of setup:
> 
> 1. TextNow's LIMITED options, and features are less applicable to you.
> 2. You gain the benefit of a centralized call log/history in 3CX (answered or not) and stable vm for missed calls
> 3. options and features of a PBX become available such as digital receptionist, call recording, multi-user, so on so forth
> 4. Contacts phonebook that is visable and accessable regardless of device (I have yet to see one in the textnow webapp other then the shotty log)
> 5. Im sure there are a pile more benefits I havent noticed.
> 
> The cons I have noticed to this kind of setup:
> 
> 1. Texting is unavailable and still requires a TextNow phone app or their webapp
> 2. Setup and configuration requirements and limitations are unknown
>    
>    * My setup consists of 3CX Self-Hosted Enterprise Edition (it was a free upgrade) installed on a Google Cloud e2-micro server with a "Lock-In" TextNow number (purchased to remove the effects of outbound inactivity)
> 3. 1 TextNow SIP Trunk account can be connected at a time (or at least from what I have found during my testing)
> 4. Various other limitations or requirements I have yet to notice.
> 
> In Short, Hell yeah you can utilize your TextNow Credentials via 3rd party PBX/voIP software.

are you still using 3cx? it is working when i am calling but i cannot receive calls with textnow SIP.

**Comment by rasterburn1** [comment]
But yeah, I havent used 3CX for about a year now, the most recent system I dropped my TextNow Creds into was a FusionPBX with similar success as I did with 3CX, same limitations (on TextNow end), same perks, features, bonuses (on PBX end) at a fraction of the price of 3CX for the same features, though it took alot of fighting and dickin around to get it to work (almost had to piggyback through the 3CX server just catch Auth. from TextNow), Though I have noticed that the TextNow Devs. seem to spend an excessive ammount of time screwing and dickin around with their server webapp code, the way it, the mobile apps, and the mobile data simcard users all connect and authenticate through the servers (which are NAT-Proxied with private network IP address'), now being you can make outbound calls but cant receive inbound calls through your 3CX which tells me you have a mis-configuration thats causing some sort of a SIP-AUTH error, I assume you are connecting via your SIP Auth. Credentials (extracted from lines and lines of TextNow connection/communication Logs gathered from Edge/Chrome/Opera/Firefox dev. console) and not the TextNow login credentials you use to log into the mobile app or the web app for texting.... yes, no, they are horribly differnt, they are not the same... also I assume you have whitelisted the appropriate IP address' / Networks that the TextNow SIP reg/auth servers use to send out notificactions, invites, registration and auth requests so on so forth... also that your firewall isnt blocking the inbound SIP requests etc. from TextNow.

**Comment by khyrfrj** [comment]
then can you please tell me what credentials i should use to be able to receive inbound calls? thanks.

**Comment by rasterburn1** [comment]
Well, It's been approximately 21 months or 1.75 years since, so I don't know how much of it (if any of it) will get you any results, but if you kindly refer to my post from back then that is ~5 up, you will find my own written instructions/steps to follow that points out any important information that is needed for sip registration, authentication, invites, and what not, but primarily and most importantly the digest algorithm, username, and realm, the matching password is in there somewhere "not shown" but hidden with encryption maybe???? also that post has a link to my TextNow sip gateway config file (xml format for drop-in convenience) that, at the time and providing the sip credentials were correct, and your firewall wasn't blocking inbound from TextNow, would provide 3CX with inbound and outbound voice calls via the TextNow Sip trunk

Here is the direct link to that post for convenience of course:
https://github.com/gornostay25/TextNowSipJsPhone/issues/1#issuecomment-1397603796

But as i said, its a bit over a year and a half old now so I dont know much of it is still relavent, but, to date, its still the best instructions to getting those critical peices of info you need, as far as im aware and to the best of my knowledge anyways.

Of course, if they are no longer relavent and dont yield you any information mentioned within them, that sucks, but atleast will know exatly what your looking for.

**Comment by khyrfrj** [comment]
> Well, It's been approximately 21 months or 1.75 years since, so I don't know how much of it (if any of it) will get you any results, but if you kindly refer to my post from back then that is ~5 up, you will find my own written instructions/steps to follow that points out any important information that is needed for sip registration, authentication, invites, and what not, but primarily and most importantly the digest algorithm, username, and realm, the matching password is in there somewhere "not shown" but hidden with encryption maybe???? also that post has a link to my TextNow sip gateway config file (xml format for drop-in convenience) that, at the time and providing the sip credentials were correct, and your firewall wasn't blocking inbound from TextNow, would provide 3CX with inbound and outbound voice calls via the TextNow Sip trunk
> 
> Here is the direct link to that post for convenience of course: [#1 (comment)](https://github.com/gornostay25/TextNowSipJsPhone/issues/1#issuecomment-1397603796)
> 
> But as i said, its a bit over a year and a half old now so I dont know much of it is still relavent, but, to date, its still the best instructions to getting those critical peices of info you need, as far as im aware and to the best of my knowledge anyways.
> 
> Of course, if they are no longer relavent and dont yield you any information mentioned within them, that sucks, but atleast will know exatly what your looking for.

[TextNowSipJsPhone](https://github.com/gornostay25/TextNowSipJsPhone) this still works but when i connect to sip on any app it doesn't work, where should i put wss://web1.prod.tncp.textnow.com:443 ? i am interested to pay if you can fix it for me. my telegram- dantepy

**Comment by rasterburn1** [comment]
I might be 1 of the few people who doesnt have telegram :smiley:, in regards to "wss://web1.prod.tncp.textnow.com:443" I dont think it is critically important as "wss://" is the web sockets layer over TLS and "web1.prod.tncp.textnow.com" is their "web proxy subdomain web1", basically just the webproxy their mobile apps and webapp does alot of communicating, but I think the most important ones you may need (and this is the registration, auth, info, invite, keepalive servers and so on) are:

UDP://prod.tncp.textnow.com:5090
udp://prod.tncp.textnow.com:5070
wss://prod.tncp.textnow.com:443
stun:ice-server.us-east-1.prod.tncp.textnow.com
turn:turn-global.prod.tncp.textnow.com

but as my 3cx gateway config shows, the only address I actually needed was having "prod.tncp.textnow.com" as the sipregister and they will all use your sip username which is your 11digit Textnow number follow by a 45digit cryptographic hash

**Comment by khyrfrj** [comment]
> I might be 1 of the few people who doesnt have telegram 😃, in regards to "wss://web1.prod.tncp.textnow.com:443" I dont think it is critically important as "wss://" is the web sockets layer over TLS and "web1.prod.tncp.textnow.com" is their "web proxy subdomain web1", basically just the webproxy their mobile apps and webapp does alot of communicating, but I think the most important ones you may need (and this is the registration, auth, info, invite, keepalive servers and so on) are:
> 
> UDP://prod.tncp.textnow.com:5090 udp://prod.tncp.textnow.com:5070 wss://prod.tncp.textnow.com:443 stun:ice-server.us-east-1.prod.tncp.textnow.com turn:turn-global.prod.tncp.textnow.com
> 
> but as my 3cx gateway config shows, the only address I actually needed was having "prod.tncp.textnow.com" as the sipregister and they will all use your sip username which is your 11digit Textnow number follow by a 45digit cryptographic hash

then please send me any contact information where i can connect you directly. As i tried all your steps and still failed so i might be doing something wrong. Thanks.

**Comment by MiniSkoot32** [comment]
Was there any progress with this? Sorry I ask 4 years later
